source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc

```diff
@@ -218,6 +224,7 @@ void GradientController::recordLatencySample(MonotonicTime rq_send_time) {
   if (inMinRTTSamplingWindow() && sample_count >= config_.minRTTAggregateRequestCount()) {
     // This sample has pushed the request count over the request count requirement for the minRTT
     // recalculation. It must now be finished.
+    synchronizer_.syncPoint("pre_minrtt_update");
     updateMinRTT();
   }
```

Comment:
Rather than double-checking the lock, I think it would be much cleaner to just wrap `updateMinRTT()` in a TryLock, require that `updateMinRTT()` call-sites hold the lock, move this if-statement into `updateMinRTT()`, and call it a day.

Only one thread needs to be doing these checks, the TryLock will ensure nobody blocks waiting on the mutex, you can bail out of the function if the conditions aren't met, and you can catch instances where the update is called without holding the lock at compile time.

I'm pretty sure this would fix the problem wholistically without having to go through the dispatcher thread nonsense. You can also verify everything via the synchronizer.

Reply:
If i understand correctly, the following is the one you described, but i think it might miss updating minRTT under some circumstance.  
Let's assume we create a separate lock `min_rtt_update_mtx_`  instead of reusing existing one. 
```
if (min_rtt_update_mtx_.TryLock()){
	updateMinRTT()
}
```

```
void updateMinRTT()
{   
    ...
	{
	    absl::MutexLock ml(&sample_mutation_mtx_);
	    if (inMinRTTSamplingWindow() &&
	        hist_sample_count(latency_sample_hist_.get()) >= config_.minRTTAggregateRequestCount()) {
	          .....
	    }
	}

	...
}

```

And let's assume `minRTTAggregateRequestCount ` is 2, and we have 2 worker threads.
- At 00:00, thread 1 increments the sample count to be 1, and acquires the `min_rtt_update_mtx_` lock, enter updateMinRTT() at 00:02, exit the updateMinRTT() at 00:04.
- At 00:01, thread 2 increments the sample count to be 2, but unable to acquire the `min_rtt_update_mtx_` lock, which has been acquired by thread 1. Thread 2 is supposed to update minRTT, but in our case it won't.

Reply:
There's nothing wrong with this. The next request will come in and update the value.

Reply:
Okay, changed to use TryLock and reused the existing mutex .

And that's a fair assumption: only if we have substantial requests, we need to update minRTT. With the TryLock and reusing the existing mutex `sample_mutation_mtx_`, it would cause some misses on updating minRTT, but it should be acceptable as we assume the subsequent request is able to update minRTT sooner or later.